### PR TITLE
Updating theme dynamically without manual refresh

### LINKED ISSUE #3613: When we change theme, we have to refresh the page to update
**Describe the bug**

When we go to settings and change the theme options and save it, we have to refresh the page so the changes can take place.

**To Reproduce**

Go to settings -> change the theme and save

**Expected behavior**

When we save and the page get refreshed, it should already update your preferences

**Desktop (please complete the following information):**

- OS: Mac OS
- Browser: Chrome 
- Version: 75.0

### PR DESCRIPTION
**Pull request by shilpi230**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
Instead of manually refreshing the page after saving the theme. This ticket fixes it and changes the theme dynamically on save. 

## Related Tickets & Documents
Closes #3613

## QA Instructions, Screenshots, Recordings
- Go to User's Settings page.
- Navigate to UX tab, select some theme.
- Theme should get updated website wide.
- [Loom Recording](https://www.loom.com/share/2503b9e5c9d5457692bfbfee861e393c)
## Added tests?

- [ ] Yes
- [x] Its a bug fix which doesn't need test case.
- [ ] I need help with writing tests

## Added to documentation?

- [ ] Docs.forem.com
- [ ] README
- [x] No documentation needed

### CODE REVIEW

**Comment by shilpi230** [comment]
Can somebody please help me in fixing the test failure. It seems some setup configuration has been missing. Netlify deployment failed.

File: app/assets/javascripts/base.js.erb [hunk 1]
```diff
@@ -1,3 +1,5 @@
+//= require jquery
+//= require jquery_ujs
```
Comment (rhymes) [inline app/assets/javascripts/base.js.erb:2]: Is there a way to do this without UJS? We purposefully don't include it as we don't use it around the website, neither jQuery

File: app/assets/javascripts/utilities/updateStyles.js [hunk 2]
```diff
@@ -0,0 +1,6 @@
+async function updateStyles(theme) {
+  const css = await fetch(`https://dev.to/assets/themes/${theme}.css`);
```
Comment (rhymes) [inline app/assets/javascripts/utilities/updateStyles.js:2]: this should take into account that this is not going to run on DEV only, so you'll have to fetch the path without the URL part: `/assets/themes`.

Also, maybe we can call this `updateTheme` so it's a bit easier to find in the future

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -318,8 +318,20 @@ def set_current_tab(current_tab = "profile")
     @tab = current_tab
   end
 
-  def config_changed?
-    params[:user].include?(:config_theme)
+  def less_than_one_day_old?(user)
```
Comment (rhymes) [inline app/controllers/users_controller.rb:321]: is this a leftover?

File: app/views/users/_ux.html.erb [hunk 4]
```diff
@@ -1,6 +1,6 @@
 <%= javascript_packs_with_chunks_tag "stickySaveFooter", defer: true %>
 
-<%= form_for @user, html: { id: "ux-customization-form", class: "sticky-footer-form grid gap-6" } do |f| %>
+<%= form_for @user, remote: true, authenticity_token: true, html: { id: "ux-customization-form", class: "sticky-footer-form grid gap-6" } do |f| %>
```
Comment (rhymes) [inline app/views/users/_ux.html.erb:3]: I think you can replace by manually using `sendFetch` on save https://github.com/forem/forem/blob/master/app/assets/javascripts/utilities/sendFetch.js

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -42,17 +42,17 @@ def update
     if @user.update(permitted_attributes(@user))
       RssReaderFetchUserWorker.perform_async(@user.id) if @user.feed_url.present?
       notice = "Your profile was successfully updated."
-      if config_changed?
-        notice = "Your config has been updated. Refresh to see all changes."
-      end
```
Comment (rhymes) [inline app/controllers/users_controller.rb:47]: this should probably stay

**Review by rhymes — CHANGES_REQUESTED** [review]
Hi @shilpi230, thanks for taking this on. I played a bit with it and it's super cool!

Aside from the inline notes I noticed that it's triggered on radio select, which makes the "save" button a little bit of a UX conundrum, maybe @ludwiczakpawel can weigh in here

The failing test is because you have removed dire noticed in the redirect that should probably stay there as it's referred to any config save

**Review by ludwiczakpawel — CHANGES_REQUESTED** [review]
For some reason, when I switch to _default_ theme, the _minimal light_ is actually being loaded.

---

Also, agree with @rhymes there's something off with UX. When you switch the theme, then go to another page without saving  you can still see that new theme. And that's a little suboptimal if you haven't saved it.

File: app/assets/javascripts/utilities/updateStyles.js (hunk 2 above)
Comment (shilpi230) [reply app/assets/javascripts/utilities/updateStyles.js:2]: sure will take this in account and makes changes! Thanks!

File: app/controllers/users_controller.rb (hunk 3 above)
Comment (shilpi230) [reply app/controllers/users_controller.rb:321]: no.

File: app/controllers/users_controller.rb (hunk 5 above)
Comment (shilpi230) [reply app/controllers/users_controller.rb:47]: @citizen428 asked me to remove this piece of code. Once manual refresh is fixed.

**Comment by shilpi230** [comment]
@ludwiczakpawel _"For some reason, when I switch to default theme, the minimal light is actually being loaded."_ i.e because I thought for default we are using minimal light theme. Which I am assuming is not correct from your statement. I will fix that. Thanks @rhymes @ludwiczakpawel for your comments. Will update those changes. Thanks!

**Comment by shilpi230** [comment]
@rhymes test case is failing in this file `spec/system/user_selects_looking_for_work_spec.rb` Which is expecting this `Your profile has been updated`  but getting something like this `DEV(local)\nWrite a post\nSetup not completed yet, missing suggested tags and suggested users. Please visit the configuration page.` Which is somehow telling that the some configuration is missing.

File: app/controllers/users_controller.rb (hunk 3 above)
Comment (rhymes) [reply app/controllers/users_controller.rb:321]: so what is it used for? I can't find it used in this PR and it's not present in master 😕

File: app/assets/javascripts/utilities/updateStyles.js (hunk 2 above)
Comment (nickytonline) [reply app/assets/javascripts/utilities/updateStyles.js:2]: As @rhymes suggested, so this should be good.

```suggestion
async function updateStyles(theme) {
  const css = await fetch(`/assets/themes/${theme}.css`);
```

File: app/views/users/_ux.html.erb [hunk 6]
```diff
@@ -62,3 +62,19 @@
     <button class="crayons-btn" type="submit">Save</button>
   </div>
 <% end %>
+<script>
+  document.getElementById("user_config_theme_default").onclick = function() {
+    updateStyles("minimal");
+  };
+  var config_themes = ["night", "minimal_light", "pink", "ten_x_hacker"];
+  for( let theme of config_themes){
```
Comment (nickytonline) [inline app/views/users/_ux.html.erb:70]: refactor (blocking): This means we're adding `n` number of onclicks for changing themes. We can do this with one `click` handler via `addEventListener` and in the event handler, we can detect which theme was clicked.

Here's some sample code.

```javascript
document.querySelector('.theme-selector-field').addEventListener('click', e => {
 const { target } = e;

 if (target.getAttribute('name') === 'user[config_theme]') {
    alert(`You clicked on the ${target.value} theme`);
 }
});
```

Also, the radio button group should be wrapped in a `<fieldset />` instead of a `<div />`, but that issue is not related to this PR.

**Review by nickytonline — CHANGES_REQUESTED** [review]
@shilpi230, thanks for this. Just some request changes.

File: app/controllers/users_controller.rb (hunk 3 above)
Comment (shilpi230) [reply app/controllers/users_controller.rb:321]: Ohh really sorry for this,  I think its leftover from conflict that I resolved. Will remove it if its not in master. Really sorry for this. Thanks!

File: app/views/users/_ux.html.erb (hunk 6 above)
Comment (nickytonline) [reply app/views/users/_ux.html.erb:70]: I created #11243 for the fieldset issue mentioned.

File: app/controllers/users_controller.rb (hunk 3 above)
Comment (rhymes) [reply app/controllers/users_controller.rb:321]: ahah no worries :) thanks for taking the time to look into this!

**Comment by cmgorton** [comment]
Hi @shilpi230 !

Thank you so much for the PR and for taking the leap and contributing to DEV! We are super happy to have your help 😄

With that said, we are in the process of going through and cleaning up some of our PR history to help us better prioritize what we need to be focusing on. Since this PR has not been active for a while I am going to go ahead and close it for now.

Thanks again for contributing! Ping me anytime if you have questions!
